Update tour events to include `editor_type`

okmttdhr opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
Related to #
-->


## Proposed Changes

Update tour events in the editor to include `editor_type` (in this case, site | post | page).

Because there are future analyses we can do where we know a user interacts more with Tour in certain cases, in the case of write flow or using the Assembler.

Part of https://github.com/Automattic/wp-calypso/issues/73142
Follow up for https://github.com/Automattic/wp-calypso/pull/73565

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Open `/site-editor/<your-site>`
- Open Welcome Guide
- Verify events with [Tracks Vigilante Chrome Extension](https://github.com/Automattic/tracks-chrome-extension)

<img width="756" alt="Screen Shot 2023-02-27 at 17 15 43" src="https://user-images.githubusercontent.com/5287479/221509802-67b89201-e662-46de-b254-cd08eaca7cc4.png">

- Open `/post/<your-site>/<site-id>`
- Open Welcome Guide
- Verify events with [Tracks Vigilante Chrome Extension](https://github.com/Automattic/tracks-chrome-extension)
<img width="763" alt="Screen Shot 2023-02-27 at 17 18 11" src="https://user-images.githubusercontent.com/5287479/221510230-4aa3cca2-910e-42d8-8cf4-ca722bf04f5a.png">

- Open `/page/<your-site>/<site-id>`
- Open Welcome Guide
- Verify events with [Tracks Vigilante Chrome Extension](https://github.com/Automattic/tracks-chrome-extension)

<img width="757" alt="Screen Shot 2023-02-27 at 17 17 29" src="https://user-images.githubusercontent.com/5287479/221510095-20396455-f391-4110-8643-ec67ba3c45a1.png">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

okmttdhr committed:
Update tour events to include `editor_type`

okmttdhr committed:
Add tests for `getEditorType`

okmttdhr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/get-editor-type.ts:
```diff
@@ -0,0 +1,17 @@
+type EditorType = 'site' | 'post' | 'page';
+export const getEditorType = (): EditorType | undefined => {
+	const editorSelector = select( 'core/editor' );
+	const postType = editorSelector.getCurrentPostType();
```
Since `getCurrentPostType` returns the post type, I assumed that if it is null, it is `site` in our case.

Docs 📝
https://developer.wordpress.org/block-editor/reference-guides/data/data-core-editor/#getcurrentposttype
https://developer.wordpress.org/reference/classes/wp_post/

and1can replied:
Hi @okmttdhr, 

I am new to the codebase.

1. How do you know that `getCurrentPostType()` yields the value `site`. Looking in https://developer.wordpress.org/reference/classes/wp_post/ and navigating to `$post_type`, there is a comment that mentions `post` and `page`, but `site` is not mentioned in that section.
2.  Are there any more docs I should read to gain context on the values of `getCurrentPostType()`?

okmttdhr replied:
Hi @and1can, thank you for your question.

> How do you know that getCurrentPostType() yields the value site

`getCurrentPostType` does **not** return `site` as a value since `getCurrentPostType` returns post type. 
Please take a look at this document. https://developer.wordpress.org/themes/basics/post-types/

With `getCurrentPostType`, we at least know if it is `post` or `page`. In this case, we wanted to distinguish `post`, `page`, and `site`, so I treated the `null` case as `site`.

miksansegundo replied:
While I think there isn't an ideal solution, it might be better to use `select( 'core/edit-site' )` to detect the Site Editor because it's already being used in the [Welcome tour](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.tsx#L61) and in [Jetpack](https://github.com/Automattic/jetpack/blob/2e56d0d24a97c096a5bf79b0fbfd2c1fb43e99ff/projects/plugins/jetpack/extensions/shared/get-editor-type.js#L13).

Here is a possible implementation:

```ts
import { select } from '@wordpress/data';

type PostType = 'page' | 'post' | string;

export const getEditorType = (): string | undefined => {
	// This method is not ideal. See https://github.com/WordPress/gutenberg/issues/46616
	if ( select( 'core/edit-site' ) ) {
		return 'site';
	}

	if ( select( 'core/edit-post' ) ) {
		// See post types in https://developer.wordpress.org/themes/basics/post-types/#default-post-types
		return select( 'core/editor' ).getCurrentPostType() as PostType;
	}

	return undefined;
};
```

Adding comments with links to identify potential issues or clarify something is good practice.

okmttdhr replied:
https://github.com/Automattic/wp-calypso/pull/73800#issuecomment-1449349064

miksansegundo commented:
I found that the property `$name` in the core PHP class [WP_Block_Editor_Context](https://github.com/WordPress/wordpress-develop/blob/031cedc06d612c4a74f7b6d3e77d8145f2954a25/src/wp-includes/class-wp-block-editor-context.php#L17) explains the different editor types:

1. `core/edit-site`: The site editor at `/wp-admin/site-editor.php`.
2. `core/edit-post`: The post editor at `/wp-admin/post.php`.
3. `core/edit-widgets`: For classic themes, the widgets editor at `/wp-admin/widgets.php`.
4. `core/customize-widgets`: For classic themes, the customizer editor at `/wp-admin/customize.php`.

The `$name` matches the name of the [data stores](https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/data) in Gutenberg. 

There are different approaches for detecting the editor type in JS:

1. Checking the URL.
2. Checking the DOM.
3. Checking the data store loaded.
Jetpack [uses](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/shared/get-editor-type.js) this approach. See the stores and selectors in [/docs/reference-guides/data](https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/data). 
   - `core/edit-site`
   - `core/edit-post`
   - `core/edit-widgets`
7. Checking for `null` on `getCurrentPostType()`. 
This PR uses this approach.

None of them seem to be ideal because they could change in the future.

There is an open [issue](https://github.com/WordPress/gutenberg/issues/46616) in Gutenberg that proposes adding `isSiteEditor` to differentiate the `Site Editor` from others. Let's follow that discussion.

miksansegundo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/test/get-editor-type.test.ts:
```diff
@@ -0,0 +1,36 @@
+		} );
+		expect( getEditorType() ).toEqual( 'page' );
+	} );
+	it( 'should return `site` when the post type is `null`', () => {
```
Kudos for adding tests 🙌 

Unfortunately, the test for `site` when the post type is `null` won't help to detect if this implementation breaks in the future if `select( 'core/editor' ).getCurrentPostType()` returns the `post_type` for templates `wp_template`.

okmttdhr replied:
Changed the implementation by [this](https://github.com/Automattic/wp-calypso/pull/73800#issuecomment-1449349064). 
I suppose we don't need tests here since we prefer to see a stable utility on the Gutenberg side. https://github.com/Automattic/wp-calypso/pull/73800/commits/3fa59ec3f58a36bf6ed44486df5b01a96068fb46

miksansegundo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.tsx:
```diff
@@ -90,6 +100,8 @@ function WelcomeTour( { siteIntent }: { siteIntent?: string } ) {
 		isSidebarOpened ||
 		( isWithinBreakpoint( '<782px' ) && ( isInserterOpened || isSettingsOpened ) );
 
+	const editorType = getEditorType();
```
Because `isSiteEditor` is already in the same [file](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.tsx#L61), it can be used to determine whether the editor type is `post` or `site` because those are the only two types of editors that show the Welcome tour.

```diff
- const editorType = getEditorType();
+ const editorType = isSiteEditor ? 'site' : 'post';
```

arthur791004 replied:
I'd prefer to keep using `getEditorType` and then we don't need to assume the Welcome tour is shown only on the site and post editor

miksansegundo reviewed (approved):
LGTM 👍 I verified the prop `editor_type` in the track events triggered by the Welcome tour when editing a site, a post, and a page.

For this PR, passing the prop `editor_type: page` in the tracks events is valuable for the stats, even though pages are edited in the `Post editor`. See the types of editors in my [comment](https://github.com/Automattic/wp-calypso/pull/73800#issuecomment-1447821788).

Please review my comments on other possible implementations..

I couldn't find the requirements for testing the `/editing-toolkit` plugin in the description. The changes in the `/editing-toolkit` plugin ([WordPress.com Editing Toolkit](https://wordpress.org/plugins/full-site-editing/)) have to be tested using the guide: PCYsg-ly5-p2
 
### Screenshots

#### Site Editor (template)
**Simple**
<img width="1043" alt="Screenshot 2566-02-28 at 14 08 37" src="https://user-images.githubusercontent.com/1881481/221834284-87dd399f-9b84-4906-b4d4-2ffc3c4fd917.png">

**Atomic**
<img width="1049" alt="Screenshot 2566-02-28 at 18 26 12" src="https://user-images.githubusercontent.com/1881481/221841948-bac3548d-bd3b-40dc-bde1-9fe5ffa00197.png">

#### Post Editor (page)
**Simple**
<img width="1051" alt="Screenshot 2566-02-28 at 14 06 49" src="https://user-images.githubusercontent.com/1881481/221834323-98c35690-e6e1-4bac-be26-7eec4076f365.png">

**Atomic**
<img width="1036" alt="Screenshot 2566-02-28 at 18 28 46" src="https://user-images.githubusercontent.com/1881481/221842140-461ccc01-fefd-4d29-b62d-225356244dfb.png">

#### Post Editor (post)
**Simple**
<img width="1051" alt="Screenshot 2566-02-28 at 14 04 48" src="https://user-images.githubusercontent.com/1881481/221834330-b86daa2e-7457-43c3-a2c6-93c62df2f4ac.png">

**Atomic**
<img width="1046" alt="Screenshot 2566-02-28 at 18 30 11" src="https://user-images.githubusercontent.com/1881481/221842224-0df6b356-d3a3-4658-aa62-1a09c294ca5c.png">

okmttdhr commented:
>Checking the URL.
Checking the DOM.

Yeah, I did not choose these because I thought they would be prone to change, and we wanted to use business logic that we could control (e.g., looking at redux states) if possible.

>The $name matches the name of the [data stores](https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/data?rgh-link-date=2023-02-28T09%3A09%3A27Z) in Gutenberg.
>Checking the data store loaded.
Jetpack [uses](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/shared/get-editor-type.js?rgh-link-date=2023-02-28T09%3A09%3A27Z) this approach. See the stores and selectors in [/docs/reference-guides/data](https://github.com/WordPress/gutenberg/tree/trunk/docs/reference-guides/data?rgh-link-date=2023-02-28T09%3A09%3A27Z).

I did not find these out! Thank you for letting me know. @miksansegundo 
I guess the method of checking for the existence of a store will work longer than relying on the `getCurrentPostType`'s implementation! I'll change it that way.

okmttdhr committed:
Check the existence of store to see editor type

Beware when using this method to figure out if we are in the site editor.
https://github.com/WordPress/gutenberg/issues/46616#issuecomment-1355301090
https://github.com/Automattic/jetpack/blob/2e56d0d/projects/plugins/jetpack/extensions/shared/get-editor-type.js

okmttdhr commented:
>https://github.com/Automattic/wp-calypso/pull/73800#pullrequestreview-1317079209
>For this PR, passing the prop editor_type: page in the tracks events is valuable for the stats, even though pages are edited in the Post editor. See the types of editors in my https://github.com/Automattic/wp-calypso/pull/73800#issuecomment-1447821788.

>https://github.com/Automattic/wp-calypso/pull/73800/files#r1119844056


I implemented it this way https://github.com/Automattic/wp-calypso/pull/73800/commits/3fa59ec3f58a36bf6ed44486df5b01a96068fb46, but let me know if we prefer to send just `post` or `site` using `isSiteEditor`.

arthur791004 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/get-editor-type.ts:
```diff
@@ -0,0 +1,45 @@
+type PostType =
+	| 'post'
+	| 'page'
+	| 'post'
```
We can remove this as it's the first one 😆

okmttdhr committed:
Remove duplicate type

arthur791004 reviewed (approved):
LGTM 👍
